### ISSUE TITLE
Issues when running multiple environments with UDP controller

### ISSUE THREAD

**Issue by HenryLHH** [esmini/esmini#559]
Hi there, 

I'm trying to run multiple environments in parallel and specify different `BasePort` of the UDP controller configuration in .xosc file. Also for each environment, I have changed their `--osi_receiver_ip` when launching `./bin/esmini`. 
However, we have an issue when opening multiple environments, where we receive the following warning for some IPs that we allocate. 
```
bind socket: Address already in use
close socket: Bad file descriptor
```
Is there a better way we can run multiple environments in parallel?

**Comment by eknabevcc** [comment]
To sort things out, 
1. With multiple environments in parallel, I assume you mean multiple esmini instances running on the same computer, right?
2. Can you run multiple instances without OSI (without specifying the --osi_receiver_ip argument)? This is expected to work.
3. It's when adding OSI for more than one instance that you get the bind socket error, right?

If the above assumptions are correct, I think it's the OSI part which cause the error. You can specify a target receiver for OSI packages, but not port - it's hard coded. So, first instance will be able to open the port for outgoing OSI messages, but not the additional instances. 

It's an interesting setup, may I ask what is your use case, briefly? That could help us to find best solution.

**Comment by HenryLHH** [comment]
Thanks for the prompt reply! What you've summarized is correct and aligned with our current issue. 
For the OSI receiver, we've already changed the receiver's IP, for example, the first receiver at `127.0.0.1`, the second at `127.0.0.2`. This goes well with fewer instances (<128 for instance), but when we ran more than 200 instances, the connection seems to be less stable, and the bind socket will then occur. 

In general, our use case is to run parallel Esmini instances for accelerated evaluation. We are resorting for different solutions right now. For example, we notice we can use the ctype package to load the dynamic libraries directly without using OSI format or UDP communication. We have a follow-up issue when modifying [this example](https://github.com/esmini/esmini/blob/949b84c0852347c436f831174594b68957dbc597/EnvironmentSimulator/code-examples/hello_world/drive.py) when changing the `SE_SimpleVehicleControlBinary` to `SE_SimpleVehicleControlAnalog`. 
We first received an error `ctype.ArgumentError: argument 2: <class 'TypeError'>: Don't know how to convert parameter 2` that indicates a wrong type of `dt = se.SE_GetSimTimeStep()`, then we manually convert it to `ctype.c_double` and conduct a similar conversion for pedal and wheel arguments. Then this error disappears, but we got `Segmentation fault (core dumped)` when executing `SE_SimpleVehicleControlAnalog`. We do not see any other examples using `SE_SimpleVehicleControlBinary` in python, could you kindly help resolve this?

**Comment by eknabevcc** [comment]
Ah, I see. You already found a good direction based on API calls, without our assistance. Great!

Just for my curiosity: Interesting and smart idea to run multiple instances on same machine with different (local) IP addresses. Since all loopback addresses (`127.*.*.*`) points to the same (local) host, do you filter out packages destined for specific IP address on the receiver/client side somehow? I guess otherwise, each client would receive all packages from all esmini instances.

But, let's focus on the API based approach. If you're in Python environment I think the ctypes approach is perfect. Also in Python environment you can receive OSI. See example: [osi_groundtruth.py](https://github.com/esmini/esmini/blob/master/EnvironmentSimulator/code-examples/hello_world/osi_groundtruth.py). This will provide most information. But the preferred approach, from performance point of view, is to fetch object states by dedicated function calls, like in this simple example: [get_states.py](https://github.com/esmini/esmini/blob/master/EnvironmentSimulator/code-examples/hello_world/get_states.py). The information you can get via API calls are a bit limited compared what's available via OSI, but in many cases enough.

Let us know if you run into issues.
